### ISSUE TITLE
Expo-image-picker extremely slow to provide result when selecting videos

### ISSUE THREAD

**Issue by ChristopherGabba** [expo/expo#26853]
### Minimal reproducible example

**[LINK TO EXPO SNACK](https://snack.expo.dev/@christophergabba/d2d3e4)**

Expo Snack Result Screenshot:
![image](https://github.com/expo/expo/assets/122328438/ea629df6-41dd-4456-b116-dc8e9f204430)

### **Expected Behavior:**
Once you select a photo or video, you get that photo or videos file path and metadata instantly, or at least relatively quickly. At a minimum, you should get the metadata instantly on the videos duration. 

### **What is Occurring:**
I've done the math and the expo-image-picker takes about 1 second for every 3.5 seconds of video duration. So if you multiply that out:

- 2 seconds for a 7 second video
- 43 seconds for a 2.5 minute video
- 85 seconds for a 5 minute video
- Etc.

Of course this is with my iPhone 12's processing power, the resolution of the video I've selected, etc. What this means is that if a user accidentally selects a 10 minute video with the picker -> BOOM the app is frozen for **2 minutes and 40 seconds** with no means whatsoever to stop it (at least not to my knowledge). 

In my application, we are returning an error if the video is longer than 30s. Using expo image picker, if the user selects a 2.5 minute video, it takes an eternity to return the meta data of the video, just so we can simply return an error that the video is longer than 30s. Obviously this makes our app look like its frozen or stuck to the user, which leads to a **terrible** user experience. 

In my application, we are also using ShareIntent, where you can share from the native iOS phone library into the app. I can take the exact same 10 minute video and share it into my app and I get the metadata instantly, where as it would take nearly 3 minutes using the ImagePicker.

Essentially in its current state, I have no idea how this image picker could be used in production to select videos-- the app appears completely frozen for long videos. At a _minimum_, there should be a callback of some sort to get the % complete of the video caching, that way we could at least tell the user what's going on. The function literally appears frozen.

### **Exact steps to reproduce:**
1. Take any video from your iPhone that is 1.5 minutes plus or just record a new one with iPhone camera so that it is on top of the photo stack
2. Use the expo snack above to open the image picker and select the long video and see how long it takes to alert.

**Note:**
I am aware of the `videoMaxDuration` property, but that only works with `allowsEditing: true`. If `allowsEditing` is turned on, it forces the user to also edit the crop of the photo, which is not what we want in our application. Maybe an alternate solution is to have two different props `allowsVideoEditing` and `allowsImageEditing` as mentioned here: [Feature Request](https://expo.canny.io/feature-requests/p/expo-image-picker-separate-allowsediting-prop-into-allowsvideoediting-and-allows)

### Environment

Reproduced in expo snack above on Expo-go in the iPhone, but this occurs in our production beta version and development builds, etc. I'm at my other job at the moment, so I don't have access to the env info at the moment. 

### Thank You
Thank you in advance for working on this, as this is an important part of my app that I would love to fix. I'll gladly provide any support necessary.

**Comment by pfcodes** [comment]
This behavior has significantly affected the UX of my app as well

**Comment by hugo1717** [comment]
I am experiencing the same behaviour. When selecting an 15 seconds video taken by an iPhone, the App freezes for 20 seconds. Any solution or alternatives?

**Comment by ChristopherGabba** [comment]
@hugo1717 Wow that's way slower than what I'm experiencing. What phone model are you using?

**Comment by hugo1717** [comment]
@ChristopherGabba, I am using an iPhone 12 on simulator and iPhone 11 on real device. The video I used was taken from the iPhone and was 30mb...

**Comment by Soul-Clinic** [comment]
Exactly the same,  clients may think it does work at all.
No matter released or debug version.

**Comment by jparkrr** [comment]
Perhaps it would work with `preferredAssetRepresentationMode: "current"`?

> A mode that uses the current representation to avoid transcoding, if possible.

Docs: https://docs.expo.dev/versions/latest/sdk/imagepicker/#uiimagepickerpreferredassetrepresentationmode

**Comment by ChristopherGabba** [comment]
@jparkrr That does speed it up and apparently fix the problem.  I tried the other two modes `compatible` and `automatic` and they are both very slow, while `current` is very fast. What exactly are the repercussions of using this `preferredAssetRepresentationMode: "current"`?

Larger file size due to no transcoding? I guess I'm confused as to why this would not be the default, especially given that it's the only mode that gives the users a reasonable experience for slightly longer vidoes.

**Comment by Soul-Clinic** [comment]
@jparkrr This is exactly the solution! Thanks!

**Comment by hirbod** [comment]
While "current" does indeed speed things up by a big magnitude, I've figured that once you edit the video metadata (adding a caption or manipulating the videos GPS information, selection once again takes forever, regardless of the preferredAssetRepresentationMode 

CC @alanjhughes 

https://github.com/expo/expo/assets/504909/4d194d9c-0155-487d-b453-08f4661f05fc

**Comment by clearly-outsane** [comment]
This fixes it on ios but what do you do on android? It still hangs considerably as file size increases

**Comment by dacay** [comment]
How about some callbacks for when the picker is dismissed? So we can show something while waiting.

**Comment by elgunhuseynli** [comment]
Please at least add some callback when video transcode begin so we can show activity indicator

**Comment by emzet93** [comment]
Is it possible to somehow use image picker in a mode, which doesn't process selected item and just return `Asset` with original file uri, like `expo-media-library` does? 🤔

With `expo-media-library` you can fetch device assets and show them almost immediately, using original file path or asset as a media source. However it doesn't provide any picker UI, like `expo-image-library` does. 

Ideally would be to have an option to use picker UI  from `expo-image-picker`, but resolve it with asset immediately, without any processing.

**Comment by TheHypnoo** [comment]
Is there any idea how to solve this? Try to show a loading UI or try to improve performance.

**Comment by jselmani** [comment]
Commenting to follow potential solutions as I'm facing the exact same problem. 

Appreciate all the advice in this thread so far.

**Comment by gavrilikhin-d** [comment]
It is also super slow, when selecting a lot (100+) images

**Comment by ice-cap0** [comment]
I can confirm getting >4s delay after selecting 2MB video on Samsung s10e

**Comment by ivaniuk7531** [comment]
I am experiencing the same issue on Android — selecting large files causes a significant delay before the picker responds.

On iOS, setting `preferredAssetRepresentationMode` to `current` helped resolve this issue, but I have encountered a different problem. When I modify any metadata or edit a media file, the response time becomes extremely long.

Has anyone found a solution to this issue?

**Comment by gavrilikhin-d** [comment]
@ivaniuk7531 I've decided to completely avoid using this lib and similar, as they create copies of assets, which is causing significant delay. You can get ph assets from `expo-media-library` instead. This is instant

**Comment by hirbod** [comment]
Just a heads up, that I’ll introduce something I call the “fast path,” which automatically kicks in when the options are suitable (no quality prop, passthrough for videos, and no allowsEditing). 

## General

As soon as we change quality, size, duration, or video quality, the media has to be manipulated, so things naturally get slower. But we can skip a lot of code if the desired output is just the original file plus some metadata, nothing else. 

base64 and EXIF can still be passed through, though they add extra overhead, especially base64 (which I discourage now that Expo SDK 53 ships with Blob, FormData, atob, and btoa). 

There are better ways to handle files, and I rarely see a real use case for base64. If the server needs it, the server can create it. Lots of slowness comes from paths that read the file and make extra copies, which we can avoid when we only need the original file.

## Metadata / EXIF changes

I wondered why metadata tweaks slow down Promise resolution so much, especially for big files. When you change any metadata in Photos (add GPS, edit date, adjust Live Photo key frame, mark as favourite, etc.), Photos doesn’t overwrite the original. Instead, it keeps the original untouched and writes a side-car blob called adjustment data that describes the change.

PHPicker is a privacy-first API, so it has to hand you a self-contained file reflecting the metadata you see in Photos. The original on disk lacks the new GPS tags; the side-car has them. So when you call loadFileRepresentation (or the picker’s internal method), Photos:

1. Reads the original video/image.
2. Re-muxes it into a fresh file, injecting the new GPS, title, or rating.
3. Streams that new file to your completion handler.

For 4K videos, that “metadata bake-in” step is basically a full re-encode/copy and can take tens of seconds, hence the slowdown.

## Solution

If transcoding is disabled (passthrough) and we have direct access to the underlying `PHAsset`, we try to copy the original/full-size video resource via `PHAssetResourceManager`. This avoids `loadFileRepresentation`, which can be noticeably slower once a user tweaks only the metadata (e.g. adjusts the capture date) because the photo service marks the asset as *adjusted* and will re-render a temporary file for us. Copying the resource bytes ourselves is dramatically faster because it just streams the already-existing file.

### Another remark

Files that live only in iCloud are always slow to resolve because they have to be downloaded first. I think the picker shows a circular download progress next to the image, but I’ll double-check. We could probably make this clearer.

CC @alanjhughes just FYI, we can discuss further

**Comment by ChristopherGabba** [comment]
Man expo hired a demon with @hirbod!! This is great info! Making it fast is obviously the best solution. If there is a chance that its going to be slow, a progresscallback like `onProgress: (percent: number) => void` would be awesome so that we as devs could at least show a loader.

Perhaps you could provide the onProgress all the time and devs could just ignore it or actually use it?

**Comment by hirbod** [comment]
I went super deep into this rabbit hole and checked how Slack, WhatsApp, and Instagram handle it. Good news: with my PR, changing metadata won’t trigger a remux anymore when you use  videoExportPreset .passthrough 

What’s still inevitable is when the video has been edited in Photos. iOS can stitch those edits in real time, but the final file only gets written in the background when the phone is idle and charging, or the first time you access it.

All the apps act the same: open a video that was just edited and it takes ages to resolve. WhatsApp just hangs, Slack shows a preview with no image until it’s ready, and Instagram spins a loader.

AFAIK there’s no progress callback, except for [iCloud downloads](https://developer.apple.com/documentation/photos/phimagerequestoptions/progresshandler?language=objc).

I think we could add support for this. For the re-mux step I doubt we’ll get real progress since the duration is unpredictable. But because we're already returning a promise, you could set a “loading” state when the selection beginns and show a spinner as long as it takes to resolve, and clear it in your `finally` block.

**Comment by ChristopherGabba** [comment]
![Image](https://github.com/user-attachments/assets/58bebec1-ba59-4c27-9d67-d12c8692a3d5)

**Comment by adshodgson** [comment]
When is this being released, this is exactly what i need right now 😅

**Comment by geoffb273** [comment]
Agreed this is necessary for me too. Would love for this to be released ASAP

**Comment by hirbod** [comment]
We're releasing this in SDK54. You can install the canary/nightly version of expo-image-picker, but there is no guarantee it works with the current expo-modules-core/ SDK 53 core.

**Comment by TheHypnoo** [comment]
Hey, but this issue also happens on Android, I understand it's only been fixed for iOS, right?

**Comment by hirbod** [comment]
That is correct @TheHypnoo. Would you mind to share a video with me regarding Android. Is it slow with multiple videos, single videos, and "how slow" is it, which format etc. Some information/ a repro could help here.

**Comment by FelipeSchiavini** [comment]
To document it here: ImagePicker.UIImagePickerPreferredAssetRepresentationMode.Current was much faster, but I had issues uploading to Cloudflare Images due to the HEIC format. on IOS Im having to use ImageManipulator to change the format to JPEG before upload it

**Comment by hirbod** [comment]
@FelipeSchiavini this is expected. I also raised this internally, as I think we need to make people more aware about the implications. Speed comes with the cost of compat issues, as we're just passing the original. The speed comes as we skip the conversion step, but I feel like doing conversion pre-upload feels much more natural instead artificially delaying a selection.
